src/z3c/recipe/compattest/runner.py

```diff
@@ -48,17 +49,22 @@ def start(self):
         )
 
     def poll(self):
-        self.exitcode = self.process.poll()
-        if self.exitcode is not None:
+        if self.process is not None:
+            self.exitcode = self.process.poll()
+        if self.exitcode is not None and self.process is not None:
             self.end = time.time()
             # We're done, get it all
             data = self.process.stdout.read()
+            # Close the pipes and cleanup.
+            self.process.communicate()
```

Comment:
Why not combine the reading/closing into one statement, namely

```python
data = self.process.communicate()[0]
```

?

Reply:
Good idea, thanks.

Reply:
Apparently [that breaks PyPy3-7.1](https://travis-ci.org/github/zopefoundation/z3c.recipe.compattest/jobs/672912967). There seems to be a buffering issue combining direct reads and the `communicate()` method.

Reply:
I don't suppose you've [filed a PyPy bug](https://bitbucket.org/pypy/pypy/issues)?

Reply:
It's already fixed; the version on Travis in outdated.

Reply:
Do you know which pypy3 version fixed that bug?

`dist: bionic` on Travis currently has pypy3 7.2.0.

Actually it doesn't matter; the workaround is simple.

Reply:
It works for me with pypy3 7.3.0.